src/dragonball/src/device_manager/vfio_dev_mgr/mod.rs

```diff
@@ -0,0 +1,805 @@
+                return Err(e);
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Attach all configured VFIO device to the virtual machine instance.
+    pub fn attach_devices(
+        &mut self,
+        ctx: &mut DeviceOpContext,
+    ) -> std::result::Result<(), StartMicroVmError> {
+        // create and attach pci root bus
+        #[cfg(feature = "host-device")]
+        if ctx.pci_hotplug_enabled {
+            let _ = self
+                .create_pci_manager(
+                    ctx.irq_manager.clone(),
+                    ctx.io_context.clone(),
+                    ctx.res_manager.clone(),
+                )
+                .map_err(StartMicroVmError::CreateVfioDevice)?;
+        }
+        for (idx, info) in self.info_list.clone().iter().enumerate() {
+            self.create_device(&info.config, ctx, idx)
+                .map_err(StartMicroVmError::CreateVfioDevice)?;
+        }
+        Ok(())
+    }
+
+    pub fn remove_device(
+        &mut self,
+        ctx: &mut DeviceOpContext,
+        hostdev_id: &str,
+    ) -> std::result::Result<(), VfioDeviceError> {
+        if !cfg!(feature = "hotplug") {
+            return Err(VfioDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        slog::info!(
+            ctx.logger(),
+            "remove VFIO device";
+            "subsystem" => "vfio_dev_mgr",
+            "hostdev_id" => hostdev_id,
+        );
+        let device_index = self
+            .get_index_of_hostdev_id(hostdev_id)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+        let mut info = self
+            .info_list
+            .remove(device_index)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+
+        self.remove_vfio_device(ctx, &mut info)
+    }
+
+    /// prepare to remove device
+    pub fn prepare_remove_device(
+        &self,
+        ctx: &DeviceOpContext,
+        hostdev_id: &str,
+        result_sender: Sender<Option<i32>>,
+    ) -> std::result::Result<(), VfioDeviceError> {
+        if !cfg!(feature = "hotplug") {
+            return Err(VfioDeviceError::UpdateNotAllowedPostBoot);
+        }
+
+        slog::info!(
+            ctx.logger(),
+            "prepare remove VFIO device";
+            "subsystem" => "vfio_dev_mgr",
+            "hostdev_id" => hostdev_id,
+        );
+
+        let device_index = self
+            .get_index_of_hostdev_id(hostdev_id)
+            .ok_or(VfioDeviceError::InvalidConfig)?;
+
+        let info = &self.info_list[device_index];
+        if let Some(dev) = info.device.as_ref() {
+            let callback: Option<Box<dyn Fn(UpcallClientResponse) + Send>> =
+                Some(Box::new(move |result| match result {
+                    UpcallClientResponse::DevMgr(response) => {
+                        if let DevMgrResponse::Other(resp) = response {
+                            if let Err(e) = result_sender.send(Some(resp.result)) {
+                                info!("send upcall result failed, due to {:?}!", e);
```

Comment:
Two sugguestions here:

1. Is the log level right? Shall we use error level?
2. Is right not to handle this error?

Reply:
1. good catch, I have changed the log level.
2. If it goes wrong here, normally means that channel is broken, which is irreparable problem and rarely happen, so we just print out the error log and find clue to fix the problem later.